### PR TITLE
[AutoDiff] remove some jvp:, vjp: from stdlib

### PR DESCRIPTION
**Pull request by marcrasi**
Removes some `jvp:` and `vjp:` from the stdlib. This is partial progress towards unblocking TF-1001.

Some bugfixes were necessary:
* `DenseMapInfo<AutoDiffConfig>` wasn't canonicalizing the `GenericSignature`, causing equivalent configs to appear distinct.
* `@derivative(of:)` typechecking was not taking the generic signature into account when comparing actual/expected differential/pullback types, causing incorrect "incorrect pullback type" diagnositcs.
* In deserialization, `MF.getIdentifier` doesn't work on special identifiers like `init`, so `@derivative(of: init)` didn't work.

Open style question: When we have a choice about whether to put a `@differentiable` attribute on the original declaration, should we? This PR leaves all the `@differentiable` attributes, but it should also work if this PR removes them.

### CODE REVIEW

File: lib/Sema/TypeCheckAttr.cpp
Comment (dan-zheng) [inline lib/Sema/TypeCheckAttr.cpp]: Note: https://github.com/apple/swift/pull/28738 upstreams `@derivative` type-checking to `master`. I can handle cherry-picking this fix to `master` after that PR is merged.

I wonder if it's possible to write a unit test that can be upstreamed with the fix?

Food for thought: after code is upstreamed, perhaps we should make changes on `master` and merge them (or let them be merged) to `tensorflow`. This saves some effort compared with committing to both `tensorflow` and `master`, since code must eventually land in `master`.

File: include/swift/AST/AutoDiff.h
Comment (dan-zheng) [inline include/swift/AST/AutoDiff.h]: For future reference: I wonder what errors arise from using `DenseMapInfo<GenericSignature>::getEmptyKey()` here?

**Review by dan-zheng — APPROVED** [review]
Nice!

If rewriting `@differentiable(jvp:vjp:)` to `@derivative` in the stdlib required compiler changes, is it possible to add unit tests for the compiler changes to ensure coverage beyond stdlib compilation?

~I wonder what issues prevented rewriting `@differentiable(jvp:vjp:)` without the compiler changes.~ Edit: I should've read the PR description more carefully, sorry - I see it mentions "incorrect pullback type" diagnostics.

**Comment by rxwei** [comment]
> When we have a choice about whether to put a `@differentiable` attribute on the original declaration, should we?

I believe we should, for libraries that are built for distribution.

File: stdlib/private/DifferentiationUnittest/DifferentiationUnittest.swift
Comment (rxwei) [inline stdlib/private/DifferentiationUnittest/DifferentiationUnittest.swift]: Note for later: We should make `@derivative` and `@transpose` imply `@usableFromInline` for `internal` declarations.

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
Some patches depend on this PR, so let's try to merge it now.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by marcrasi** [comment]
Good idea about the unit tests. I'll write some and send a followup PR, so that I don't block merging this.

**Comment by marcrasi** [comment]
> > When we have a choice about whether to put a `@differentiable` attribute on the original declaration, should we?
> 
> I believe we should, for libraries that are built for distribution.

Sounds good to me. Therefore, I will keep the `@differentiable` attributes while removing vjp:/jvp:, with one exception: I'll remove `@differentiable` from all non-differentiation-related stdlib functions, because we want to avoid modifying any existing stdlib files while upstreaming.

**Pull request merged**
